File: src/main/java/fi/quanfoxes/lexer/ParenthesisType.java
package fi.quanfoxes.lexer;

import java.util.HashMap;
import java.util.Map;

public enum ParenthesisType {
    PARENTHESIS('(', ')'),
    BRACKETS('[', ']'),
    CURLY_BRACKETS('{', '}');

    private static Map<Character, ParenthesisType> map = new HashMap<>();

    static {
        for (ParenthesisType operator : ParenthesisType.values()) {
            map.put(operator.opening, operator);
        }
    }

    private char opening;
    private char closing;

    private ParenthesisType(char opening, char closing) {
        this.opening = opening;
        this.closing = closing;
    }

    public static ParenthesisType get(Character opening) {
        return map.get(opening);
    }

    public static boolean has(Character opening) {
        return map.containsKey(opening);
    }

    public Character getOpening() {
        return opening;
    }

    public Character getClosing() {
        return closing;
    }
}

File: Zigzag/Lexer/ParenthesisType.cs
using System;
using System.Collections.Generic;

public class ParenthesisType
{
	public static readonly ParenthesisType PARENTHESIS = new ParenthesisType('(', ')');
	public static readonly ParenthesisType BRACKETS = new ParenthesisType('[', ']');
	public static readonly ParenthesisType CURLY_BRACKETS = new ParenthesisType('{', '}');

<<<<<<< HEAD
	private static Dictionary<char, ParenthesisType> Map { get; } = new Dictionary<char, ParenthesisType>();
=======
	private static readonly Dictionary<char, ParenthesisType> Map = new Dictionary<char, ParenthesisType>();
>>>>>>> ec8e325... Improved code quality and implemented basic support for operator overloading

	static ParenthesisType()
	{
		Map.Add(PARENTHESIS.Opening, PARENTHESIS);
		Map.Add(BRACKETS.Opening, BRACKETS);
		Map.Add(CURLY_BRACKETS.Opening, CURLY_BRACKETS);
	}

	public char Opening { get; private set; }
	public char Closing { get; private set; }

	private ParenthesisType(char opening, char closing)
	{
		Opening = opening;
		Closing = closing;
	}

	public static ParenthesisType Get(char opening)
	{
		return Map[opening];
	}

	public static bool Has(char opening)
	{
		return Map.ContainsKey(opening);
	}

	public override bool Equals(object? obj)
	{
		return obj is ParenthesisType type &&
				base.Equals(obj) &&
			   	Opening == type.Opening &&
			   	Closing == type.Closing;
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(Opening, Closing);
	}
}